File: javaScript/gasStation.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function(gas, cost) {
    const n = gas.length;
    
    // Calculate total gas and total cost
    let totalGas = 0;
    let totalCost = 0;
    for (let i = 0; i < n; i++) {
        totalGas += gas[i];
        totalCost += cost[i];
    }
    
    // If total gas is less than total cost, impossible
    if (totalGas < totalCost) {
        return -1;
    }
    
    // Difference array and prefix sum
    const diff = new Array(n);
    for (let i = 0; i < n; i++) {
        diff[i] = gas[i] - cost[i];
    }
    
    // Find minimum prefix sum position
    let minPrefix = 0;
    let currentSum = 0;
    let minIndex = 0;
    
    for (let i = 0; i < n; i++) {
        currentSum += diff[i];
        if (currentSum < minPrefix) {
            minPrefix = currentSum;
            minIndex = i + 1;
        }
    }
    
    return minIndex % n;
};

File: java/gasStation.java
package java;

public class gasStation {
     public int canCompleteCircuit(int[] gas, int[] cost) {
     int total = 0;
        int current = 0;
        int start = 0;
        int n = gas.length;
        
        for (int i = 0; i < n; i++) {
            total += gas[i] - cost[i];
            current += gas[i] - cost[i];
            
            if (current < 0) {
                start = i + 1;
                current = 0;
            }
        }
        
        return total >= 0 ? start : -1;
    }
}
